middleware/index.js

```diff
@@ -0,0 +1,227 @@
+'use strict';
+var chalk = require('chalk');
+var yeoman = require('yeoman-generator');
+
+var wsModels = require('loopback-workspace').models;
+
+var actions = require('../lib/actions');
+var helpers = require('../lib/helpers');
+var validateName = helpers.validateName;
+var async = require('async');
+
+function listChoices(items) {
+  var choices = [];
+  items.forEach(function(item, index) {
+    choices.push({
+      name: (index + 1) + '. ' + item,
+      value: item
+    });
+  });
+  return choices;
+}
+
+module.exports = yeoman.generators.Base.extend({
+  // NOTE(bajtos)
+  // This generator does not track file changes via yeoman,
+  // as loopback-workspace is editing (modifying) files when
+  // saving project changes.
+
+  loadProject: actions.loadProject,
+
+  constructor: function() {
+    yeoman.generators.Base.apply(this, arguments);
+
+    this.argument('name', {
+      desc: 'Name of the middleware to create.',
+      required: false,
+      type: String
+    });
+  },
+
+  help: function() {
+    return helpers.customHelp(this);
+  },
+
+  loadPhases: function() {
+    var done = this.async();
+    wsModels.Middleware.find(function(err, list) {
+      if (err) {
+        return done(err);
+      }
+
+      var phases = {};
+      list.forEach(function(c) {
+        if (c.phase) {
+          phases[c.phase] = c.phase;
+        }
+      });
+
+      this.phases = Object.keys(phases);
```

Comment:
IIUC, this assumes that `wsModels.Middleware.find` returns the middleware entries sorted by the phase index in the middleware file. Is that a correct assumption to make?